[quant][pt2e] Move batch norm op between eval/train for cuda

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123957

Summary: Before in `move_exported_model_to_train/eval`, we only
switched the CPU versions of the batch norm op. This commit adds
support for the cuda versions of the op too. Note that this fix
is temporary; we won't have to differentiate between these two
cases once we have batch norm consolidation.

Test Plan:
python test/test_quantization.py -k test_move_exported_model_bn

Reviewers: jerryzh168

Subscribers: jerryzh168, leslie-fang-intel, supriyar

Differential Revision: [D56070054](https://our.internmc.facebook.com/intern/diff/D56070054)

andrewor14 committed:
[quant][pt2e] Move batch norm op between eval/train for cuda

Summary: Before in `move_exported_model_to_train/eval`, we only
switched the CPU versions of the batch norm op. This commit adds
support for the cuda versions of the op too. Note that this fix
is temporary; we won't have to differentiate between these two
cases once we have batch norm consolidation.

Test Plan:
python test/test_quantization.py -k test_move_exported_model_bn

Reviewers: jerryzh168

Subscribers: jerryzh168, leslie-fang-intel, supriyar

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56070054).

andrewor14 committed:
Update on "[quant][pt2e] Move batch norm op between eval/train for cuda"


Summary: Before in `move_exported_model_to_train/eval`, we only
switched the CPU versions of the batch norm op. This commit adds
support for the cuda versions of the op too. Note that this fix
is temporary; we won't have to differentiate between these two
cases once we have batch norm consolidation.

Test Plan:
python test/test_quantization.py -k test_move_exported_model_bn

Reviewers: jerryzh168

Subscribers: jerryzh168, leslie-fang-intel, supriyar

Differential Revision: [D56070054](https://our.internmc.facebook.com/intern/diff/D56070054)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56070054).

andrewor14 committed:
Update on "[quant][pt2e] Move batch norm op between eval/train for cuda"


Summary: Before in `move_exported_model_to_train/eval`, we only
switched the CPU versions of the batch norm op. This commit adds
support for the cuda versions of the op too. Note that this fix
is temporary; we won't have to differentiate between these two
cases once we have batch norm consolidation.

Test Plan:
python test/test_quantization.py -k test_move_exported_model_bn

Reviewers: jerryzh168

Subscribers: jerryzh168, leslie-fang-intel, supriyar

Differential Revision: [D56070054](https://our.internmc.facebook.com/intern/diff/D56070054)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56070054).

jerryzh168 commented on torch/ao/quantization/pt2e/export_utils.py:
```diff
@@ -136,20 +136,29 @@ def bn_eval(
+    # available, but right now there's no other way to differentiate between
+    # the different bn ops. Once batch norm consolidation lands, we won't have
+    # to do this anymore: https://github.com/pytorch/pytorch/pull/119496.
+    is_cuda = torch.cuda.is_available()
```
right, can we get this from the model itself?

jerryzh168 replied:
e.g. with this: https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/fx/utils.py#L225

andrewor14 replied:
oh this is great, I'll update with this util

andrewor14 committed:
Update on "[quant][pt2e] Move batch norm op between eval/train for cuda"


Summary: Before in `move_exported_model_to_train/eval`, we only
switched the CPU versions of the batch norm op. This commit adds
support for the cuda versions of the op too. Note that this fix
is temporary; we won't have to differentiate between these two
cases once we have batch norm consolidation.

Test Plan:
python test/test_quantization.py -k test_move_exported_model_bn

Reviewers: jerryzh168

Subscribers: jerryzh168, leslie-fang-intel, supriyar

Differential Revision: [D56070054](https://our.internmc.facebook.com/intern/diff/D56070054)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56070054).

jerryzh168 commented:
@andrewor14 do you still want to land this? if so please fix the CI failures

andrewor14 committed:
Update on "[quant][pt2e] Move batch norm op between eval/train for cuda"


Summary: Before in `move_exported_model_to_train/eval`, we only
switched the CPU versions of the batch norm op. This commit adds
support for the cuda versions of the op too. Note that this fix
is temporary; we won't have to differentiate between these two
cases once we have batch norm consolidation.

Test Plan:
python test/test_quantization.py -k test_move_exported_model_bn

Reviewers: jerryzh168

Subscribers: jerryzh168, leslie-fang-intel, supriyar

Differential Revision: [D56070054](https://our.internmc.facebook.com/intern/diff/D56070054)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56070054).

jerryzh168 reviewed (approved):
looks good, had a comment inline

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1840,29 +1852,33 @@ def __init__(self):
 
-        example_inputs = (torch.randn(1, 3, 3, 3),)
-        m = M().train()
+        is_cuda = torch.cuda.is_available()
```
sorry this is in test, so it's fine

andrewor14 commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot revert -m "reverting to check if it will fix rocm jobs on main"

jeanschmidt commented:
@pytorchbot revert -m "reverting to check if it will fix rocm jobs on main" -c nosignal

andrewor14 committed:
Update on "[quant][pt2e] Move batch norm op between eval/train for cuda"


Summary: Before in `move_exported_model_to_train/eval`, we only
switched the CPU versions of the batch norm op. This commit adds
support for the cuda versions of the op too. Note that this fix
is temporary; we won't have to differentiate between these two
cases once we have batch norm consolidation.

Test Plan:
python test/test_quantization.py -k test_move_exported_model_bn

Reviewers: jerryzh168

Subscribers: jerryzh168, leslie-fang-intel, supriyar

Differential Revision: [D56070054](https://our.internmc.facebook.com/intern/diff/D56070054)

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2e] Move batch norm op between eval/train for cuda"


Summary: Before in `move_exported_model_to_train/eval`, we only
switched the CPU versions of the batch norm op. This commit adds
support for the cuda versions of the op too. Note that this fix
is temporary; we won't have to differentiate between these two
cases once we have batch norm consolidation.

Test Plan:
python test/test_quantization.py -k test_move_exported_model_bn

Reviewers: jerryzh168

Subscribers: jerryzh168, leslie-fang-intel, supriyar

Differential Revision: [D56070054](https://our.internmc.facebook.com/intern/diff/D56070054)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56070054).

andrewor14 commented:
@pytorchbot merge
